[WIP] initial fixpoint iteration

carljm opened the pull request:
Pulls in a Salsa version that supports fixpoint iteration, for experimentation.

MichaReiser commented:
Nice on bringing down the regression. I'm still somewhat surprised that we see an 11% perf regression and a 4 perf regression in the salsa benchmarks (that don't have any cycles).

carljm commented:
Yeah I planned to do some more looking at profiles; thanks for doing that!

MichaReiser commented on crates/red_knot_project/tests/check.rs:
```diff
@@ -279,23 +279,4 @@ impl SourceOrderVisitor<'_> for PullTypesVisitor<'_> {
-    ("crates/ruff_linter/resources/test/fixtures/pyflakes/F401_12.py", false, true),
-    ("crates/ruff_linter/resources/test/fixtures/pyflakes/F401_14.py", false, true),
-];
+const KNOWN_FAILURES: &[(&str, bool, bool)] = &[];
```
Should we delete this functionality now that all known failures are fixed? Do we forsee situations where it is okay for a PR to introduce a new panic case over fixing the panic?

carljm replied:
It's a good question. I thought we could keep it for now. I _think_ any new cycle cases we run into should be easy to fix by adding cycle handling for a new query, but it's possible we could run into such a case where it's difficult to get the cycle handling correct and it makes more sense to split it into a separate PR, in which scenario we might still use this? I don't feel strongly.

MichaReiser replied:
That makes sense. Although that also likely means that we would start regressing on projects that Red Knot supported before. So maybe the correct order would be to fix the cycle first, and then land the behavior change?

I don't feel strongly either and am happy to go either way. It just feels nice if regressing panics is *not easy*

carljm replied:
> we would start regressing on projects that Red Knot supported before

Maybe; it depends whether the PR is actually regressing on existing files (in this case I would be more hesitant to allow using `KNOWN_FAILURES`), or if it is e.g. an unrelated Ruff linter PR adding a new test file, which reveals an existing-but-not-previously-tested cycle. This latter case is I think the strongest reason to keep `KNOWN_FAILURES` around for now.

MichaReiser reviewed (approved):
This looks good to me. Do you think it would be useful to have any documentation on when/where to add cycle fallbacks or documentation on the cycle fallback function itself?

carljm commented:
> Do you think it would be useful to have any documentation on when/where to add cycle fallbacks or documentation on the cycle fallback function itself?

Yes, I can add that. I think the latter is probably the most useful and will serve as the former by example, too? Not sure where to put the former that it would be discoverable when needed. Also, Salsa's panic message if you do hit a cycle on a query without cycle handling is pretty informative and would be enough to find existing examples to copy.

carljm commented:
Changed my mind -- too many cases of cycle handling, felt too redundant to repeat the same docs on each. So I added a paragraph to the doc comment at the top of `infer.rs`, where all but one of our queries with cycle handling live.
